gobblin-core-base/src/main/java/gobblin/writer/AsyncWriterManager.java

```diff
@@ -0,0 +1,519 @@
+   *
+   */
+  private boolean isFailureFatal()
+  {
+    return (failureAllowance == 0.0);
+  }
+
+
+  private void makeNextWriteThrow(Throwable t)
+  {
+    cachedWriteException = t;
+  }
+
+  private void maybeThrow() throws IOException {
+    if (cachedWriteException != null)
+    {
+      throw new IOException("Irrecoverable failure on async write", cachedWriteException);
+    }
+  }
+
+  private void attemptWrite(final Attempt attempt)
+  {
+    final long startTime = System.nanoTime();
+    this.recordsAttempted.mark();
+    this.asyncDataWriter.write(attempt.record, new WriteCallback() {
+
+      @Override
+      public void onSuccess(WriteResponse writeResponse) {
+
+        recordsSuccess.mark();
+        if (writeResponse.bytesWritten() > 0)
+        {
+          bytesWritten.mark(writeResponse.bytesWritten());
+        }
+        if (dataWriterTimer.isPresent()) {
+          dataWriterTimer.get().update(System.nanoTime() - startTime, TimeUnit.NANOSECONDS);
+        }
+      }
+
+      @Override
+      public void onFailure(Throwable exception) {
+        if (dataWriterTimer.isPresent()) {
+          dataWriterTimer.get().update(System.nanoTime() - startTime, TimeUnit.NANOSECONDS);
+        }
+        if (attempt.attemptNum <= numRetries) { // attempts must == numRetries + 1
+          attempt.incAttempt();
+          attempt.setPrevAttemptFailure(exception);
+          retryQueue.get().add(attempt);
+        }
+        else
+        {
+          recordsFailed.mark();
+          log.debug("Failed to write record : " + attempt.getRecord().toString(), exception);
+          // If this failure is fata, set the writer to throw an exception at this point
+          if (isFailureFatal())
+          {
+            makeNextWriteThrow(exception);
+          }
+        }
+      }
+    });
+  }
+
+  private class RetryRunner implements Runnable {
+
+    private final LinkedBlockingQueue<Attempt> _retryQueue;
+
+    public RetryRunner() {
+      Preconditions.checkArgument(retryQueue.isPresent(), "RetryQueue must be present for RetryRunner");
+      _retryQueue = retryQueue.get();
+    }
+
+    @Override
+    public void run() {
+      while (true)
+      {
+        try {
+          Attempt attempt = _retryQueue.take();
+          if (attempt != null) {
+            attemptWrite(attempt);
+          }
+        } catch (InterruptedException e) {
+          log.info("Retry thread interrupted... will exit");
+          Throwables.propagate(e);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void cleanup()
+      throws IOException {
+    // legacy api ...
+  }
+
+  @Override
+  public long recordsWritten() {
+    return recordsSuccess.getCount();
+  }
+
+  @Override
+  public long bytesWritten()
+      throws IOException {
+    return bytesWritten.getCount();
+  }
+
+  @Override
+  public void close()
+      throws IOException {
+    log.debug("Close called");
+    this.closer.close();
```

Comment:
I don't see this being done in other writers: e.g. 
gobblin-core/src/main/java/gobblin/writer/JdbcWriter.java
gobblin-core/src/main/java/gobblin/writer/FsDataWriter.java

Are you concerned about this particular class because it is interfacing with async writers?